### PR TITLE
cpu/saml21: enable 48mhz clock for usbdev

### LINKED ISSUE #12077: usbdev support for SAML21
I would like to use periph_usbdev on SAML21 based boards. Given that this is a SAM0 variant, i have naively activated support like this:

```
diff --git a/boards/saml21-xpro/Makefile.features b/boards/saml21-xpro/Makefile.features
index 53cbc6967..4e062523a 100644
--- a/boards/saml21-xpro/Makefile.features
+++ b/boards/saml21-xpro/Makefile.features
@@ -6,6 +6,7 @@ FEATURES_PROVIDED += periph_rtt
 FEATURES_PROVIDED += periph_spi
 FEATURES_PROVIDED += periph_timer
 FEATURES_PROVIDED += periph_uart
+FEATURES_PROVIDED += periph_usbdev
 
 # Put other features for this board (in alphabetical order)
 FEATURES_PROVIDED += riotboot
diff --git a/boards/saml21-xpro/include/periph_conf.h b/boards/saml21-xpro/include/periph_conf.h
index 95c4b8b2e..353857da0 100644
--- a/boards/saml21-xpro/include/periph_conf.h
+++ b/boards/saml21-xpro/include/periph_conf.h
@@ -172,6 +172,19 @@ static const adc_conf_chan_t adc_channels[] = {
 #define ADC_0_REF_DEFAULT                  ADC_REFCTRL_REFSEL_INTVCC2
 /** @} */
 
+/**
+ * @name USB peripheral configuration
+ * @{
+ */
+static const sam0_common_usb_config_t sam_usbdev_config[] = {
+    {
+        .dm     = GPIO_PIN(PA, 24),
+        .dp     = GPIO_PIN(PA, 25),
+        .d_mux  = GPIO_MUX_G,
+        .device = &USB->DEVICE,
+    }
+};
+/** @} */
 #ifdef __cplusplus
 }
 #endif
```

Unfortunately, this is not enough - I guess there is some SAML21 related quirk in the driver that needs to be ironed out. What I can see is that on the board, the _usbdev_set function is called for USBOPT_ADDRESS, but with the invalid address 0.

Is there anyone that made it work already, or that is willing to share some suggestion about how to debug the driver?

Thanks a lot,

Antonio

### PR DESCRIPTION
**Pull request by ant9000**
### Contribution description

This patch for SAML21 boards enables the 48MHz clock used by the usbdev driver, and the adds the driver configuration for a couple of boards (saml21-xpro, samr30-xpro).

### Testing procedure

I've tested a SAML21 board with both example/usbus_minimal and tests/usbus_cdc_ecm, and it works correctly. Also tested on samr34-xpro using PR #11250.

### Issues/PRs references

Closes #12077. Code adapted from #10075 as suggested by @dylad.

### CODE REVIEW

File: cpu/sam0_common/periph/usbdev.c [hunk 1]
```diff
@@ -242,8 +242,41 @@ static inline void _poweron(void)
...
+                             OSCCTRL_DFLLCTRL_CCDIS |
+                             OSCCTRL_DFLLCTRL_ENABLE;
+
+    /* Ensure COARSE and FINE are locked */
+    while((!(OSCCTRL->STATUS.bit.DFLLLCKC)) && (!(OSCCTRL->STATUS.bit.DFLLLCKF))) {}
+    while (!(OSCCTRL->STATUS.bit.DFLLRDY)) {}
+
+    /* set clock 1 to 48 MHz DFLL */
+    GCLK->GENCTRL[1].reg = GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M;
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_GENCTRL(1)) {}
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c]: I'd much rather see that DFLL setup code in `saml21/cpu.c`.
It isn't related to usbdev - that just happens to use that clock.

If we want to use the DFLL for something else in the future, we better not duplicate the setup code.

samd5x does that too - set up the clock in `cpu_init()` if `MODULE_PERIPH_USBDEV` is used, then the user (usbdev) just connects to the existing GCLK. This also has the benefit that all GCLKs are set up in one place.

Comment (ant9000) [reply cpu/sam0_common/periph/usbdev.c]: My use case for keeping the clock initialization separate is a energy harvesting node. I would like to carry out configuration via USB, but the 48MHz clock should stay off unless needed, in order to save power.

Comment (benpicco) [reply cpu/sam0_common/periph/usbdev.c]: You can connect all the peripheral to a separate GCLK that stays constant, then you can change the main clock independently.

But yes, the whole GCLK system needs some proper overhaul in RIOT.

Something like `gclk_connect()` and `gclk_disconnect()` and then disabling the clock of there are no more users should definitely be part of it.

**Comment by dylad** [comment]
@ant9000 with #10075 finally merged, do you think you could rebase and update this PR ? I would love to see USB support on SAML21.
If you don't plan to work on it anymore, do you mind if I force-push your branch to update it so we can finally move forward ?

**Comment by ant9000** [comment]
I've rebased the PR - let me know if you think something more could be done. I would also love to have official USB support on L21.

File: cpu/sam0_common/periph/usbdev.c [hunk 2]
```diff
@@ -246,6 +246,45 @@ static inline void _poweron(sam0_common_usb_t *dev)
...
+
+    /* use it as USB clock */
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK1;
+#elif defined(CPU_FAM_SAMD5X)
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK6;
```
Comment (dylad) [inline cpu/sam0_common/periph/usbdev.c]: This whole part should be rewrite with #10075 merged, the DFLL should already be used to source the CPU clock I think we should just connect the USB peripheral to GCLK0.
Further improvements could be made later.

Comment (benpicco) [reply cpu/sam0_common/periph/usbdev.c]: You probably want to introduce a separate `SAM0_GCLK_48MHZ` to which the DFLL is connected.
So you can still run the CPU at 12 MHz if you want.

**Comment by benpicco** [comment]
I think as for CPU changes, this should be enough:

```patch
diff --git a/cpu/saml21/cpu.c b/cpu/saml21/cpu.c
index a8ca51ad65..cd524a173d 100644
--- a/cpu/saml21/cpu.c
+++ b/cpu/saml21/cpu.c
@@ -66,8 +66,11 @@ static void _xosc32k_setup(void)
 
 void sam0_gclk_enable(uint8_t id)
 {
-    (void) id;
-    /* clocks are always running */
+    switch (id) {
+    case SAM0_GCLK_48MHZ:
+        _gclk_setup(SAM0_GCLK_48MHZ, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M);
+        break;
+    }
 }
 
 uint32_t sam0_gclk_freq(uint8_t id)
@@ -79,6 +82,8 @@ uint32_t sam0_gclk_freq(uint8_t id)
         return 8000000;
     case SAM0_GCLK_32KHZ:
         return 32768;
+    case SAM0_GCLK_48MHZ:
+        return 48000000;
     default:
         return 0;
     }
diff --git a/cpu/saml21/include/periph_cpu.h b/cpu/saml21/include/periph_cpu.h
index 485e6ee363..66558471a0 100644
--- a/cpu/saml21/include/periph_cpu.h
+++ b/cpu/saml21/include/periph_cpu.h
@@ -39,6 +39,7 @@ enum {
     SAM0_GCLK_MAIN  = 0,                 /**< Main clock */
     SAM0_GCLK_8MHZ  = 1,                 /**< 8MHz clock */
     SAM0_GCLK_32KHZ = 2,                 /**< 32 kHz clock */
+    SAM0_GCLK_48MHZ = 3,                 /**< 48MHz clock */
 };
 /** @} */
 ```

Then make sure to add `.gclk_src = SAM0_GCLK_48MHZ` to your `sam_usbdev_config` structs and it should just work already.

**Comment by dylad** [comment]
I allow myself to force-push your branch to update this PR. Of course, I kept your commit authorship.
@aabadie May I ask you to test this PR on SAMR34 if you can please ? I'd like to merge this PR before the hard feature freeze !

File: cpu/samd21/periph/pm.c [hunk 3]
```diff
@@ -61,23 +61,27 @@ void pm_set(unsigned mode)
...
              * Potential Wake Up sources: Synchronous (APB), asynchronous
              */
+             DEBUG_PUTS("pm_set(): setting IDLE1 mode.");
             PM->SLEEP.reg = SYSTEM_SLEEPMODE_IDLE_1;
             break;
         case SAMD21_PM_IDLE_0:
             /* Sleep mode Idle 0
              * Potential Wake Up sources: Synchronous (APB, AHB), asynchronous
             */
+            DEBUG_PUTS("pm_set(): setting IDLE0 mode.");
```
Comment (benpicco) [inline cpu/samd21/periph/pm.c:84]: Seems unrelated

File: cpu/sam0_common/periph/usbdev.c [hunk 4]
```diff
@@ -246,6 +246,12 @@ static inline void _poweron(sam0_common_usb_t *dev)
     GCLK->CLKCTRL.reg = GCLK_CLKCTRL_CLKEN
                       | GCLK_CLKCTRL_GEN(dev->config->gclk_src)
                       | GCLK_CLKCTRL_ID(USB_GCLK_ID);
+#elif defined(CPU_FAM_SAML21)
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN(dev->config->gclk_src);
+#elif defined(CPU_FAM_SAMD5X)
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK6;
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:254]: Why?

File: cpu/sam0_common/periph/usbdev.c [hunk 5]
```diff
@@ -246,6 +246,12 @@ static inline void _poweron(sam0_common_usb_t *dev)
     GCLK->CLKCTRL.reg = GCLK_CLKCTRL_CLKEN
                       | GCLK_CLKCTRL_GEN(dev->config->gclk_src)
                       | GCLK_CLKCTRL_ID(USB_GCLK_ID);
+#elif defined(CPU_FAM_SAML21)
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN(dev->config->gclk_src);
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:251]: This is already caught by the `#else` branch, no need to introduce separate branches for saml21 and samd5x

File: cpu/samd21/periph/pm.c (hunk 3 above)
Comment (dylad) [reply cpu/samd21/periph/pm.c]: Removed

File: cpu/sam0_common/periph/usbdev.c (hunk 4 above)
Comment (dylad) [reply cpu/sam0_common/periph/usbdev.c]: I forgot to remove this

Comment (dylad) [reply cpu/sam0_common/periph/usbdev.c]: Done

File: cpu/sam0_common/periph/usbdev.c (hunk 5 above)
Comment (dylad) [reply cpu/sam0_common/periph/usbdev.c]: Fixed

**Comment by dylad** [comment]
ping @aabadie

**Comment by benpicco** [comment]
Uh somehow we dropped this, but the code looks good.
Can you rebase.

**Comment by dylad** [comment]
thanks for rebase @ant9000 
@benpicco this PR should be good.

**Comment by dylad** [comment]
Thanks for your patience ! Finally SAML21 deserves some love !
